python_modules/dagster/dagster/components/core/component_hierarchy.py

```diff
@@ -0,0 +1,206 @@
+    children: Mapping[Path, ComponentDeclaration]
+    yaml_file: Optional[ComponentYamlFile] = None
+
+    @classmethod
+    def build(cls, context: ComponentLoadContext) -> "FolderComponentDeclaration":
+        found: dict[Path, ComponentDeclaration] = {}
+        for subpath in context.path.iterdir():
+            sub_ctx = context.for_path(subpath)
+            with use_component_load_context(sub_ctx):
+                component = build_component_declaration(sub_ctx)
+                if component:
+                    found[subpath] = component
+
+        component_yaml_path = context.path / "component.yaml"
+        return cls(
+            path=context.path,
+            children=found,
+            yaml_file=None
+            if not component_yaml_path.exists()
+            else ComponentYamlFile(
+                path=component_yaml_path, sha=compute_file_hash(component_yaml_path)
+            ),
+        )
+
+
+@record
+class ComponentHierarchy:
+    root: ComponentDeclaration
+
+
+def build_component_declaration(
+    context: ComponentLoadContext,
+) -> Optional[ComponentDeclaration]:
+    # Check for yaml component
+    component_yaml_path = context.path / "component.yaml"
+    if component_yaml_path.exists():
+        return YamlComponentDeclaration.build(context)
+
+    # Check for pythonic component
+    if (context.path / "component.py").exists():
+        return PythonicComponentDeclaration.build(context)
+
+    # Check for defs component
+    if (context.path / "definitions.py").exists():
+        return DefsComponentDeclaration.build(context)
+
+    # Check for python module
+    if context.path.suffix == ".py":
+        return PythonModuleComponentDeclaration.build(context)
+
+    # Check for folder component
+    if context.path.is_dir():
+        return FolderComponentDeclaration.build(context)
+
+    return None
+
+
+def build_component_hierarchy(
+    context: ComponentLoadContext,
+) -> ComponentHierarchy:
+    """Builds a hierarchy of components starting from the given context.
+
+    Args:
+        context (ComponentLoadContext): The context from which to start building the hierarchy.
+
+    Returns:
+        ComponentHierarchy: The root node of the component hierarchy.
+    """
+    root_decl = build_component_declaration(context)
+    if not root_decl:
+        raise ValueError("Expected root_decl to be defined, but got None")
+    return ComponentHierarchy(root=root_decl)
+
+
+def build_root_component(
+    context: ComponentLoadContext, hierarchy: ComponentHierarchy
+) -> Optional[Component]:
+    """Builds the root component from the given context.
+
+    Args:
+        context (ComponentLoadContext): The context from which to start building the component.
+
+    Returns:
+        Optional[Component]: The root component if found, otherwise None.
+    """
+    with use_component_load_context(context):
+        return build_component_from_node(context, hierarchy.root)
```

Comment:
something like 

`hierarchy.root.load(context)`

which would call

```
class FolderComponentDeclaration:

   def load(self, context: ComponentLoadContext):
       return DefsFolderComponent(
            path=context.path,
            children={
                path: child_decl.load(context.for_path(path)
                for path, child_decl in decl.children.items()
            },
            asset_post_processors=None,
       )
```

and `build_component_from_node` would go away 

`load` / `build` / `instantiate` / `create` whatever